### PR TITLE
[#9450] port twisted.internet.address to attrs

### PR DESCRIPTION
**Pull request by hawkowl**
https://twistedmatrix.com/trac/ticket/9450#ticket

### CODE REVIEW

File: src/twisted/internet/address.py [hunk 1]
```diff
@@ -7,79 +7,60 @@
...
-
-
-    def __repr__(self):
-        return '%s(%s, %r, %d)' % (
-            self.__class__.__name__, self.type, self.host, self.port)
-
-
-    def __hash__(self):
-        return hash((self.type, self.host, self.port))
+_IPAddress = attr.make_class("_IPAddress", {
```
Comment (hynek) [inline src/twisted/internet/address.py]: Why make_class and not just define the class? To avoid having to write a docstring? :)

Comment (hynek) [reply src/twisted/internet/address.py]: Also, does it make sense to even have _IPAddress if you’re using attrs? You’re saving two lines in each class definition but have scattered information.

File: src/twisted/internet/address.py [hunk 2]
```diff
@@ -117,66 +95,32 @@ class HostnameAddress(FancyEqMixin, object):
...
 
     if getattr(os.path, 'samefile', None) is not None:
         def __eq__(self, other):
             """
-            Overriding C{FancyEqMixin} to ensure the os level samefile
+            Overriding C{attrs} to ensure the os level samefile
             check is done if the name attributes do not match.
             """
-            res = super(UNIXAddress, self).__eq__(other)
+            res = self.name == other.name
```
Comment (hynek) [inline src/twisted/internet/address.py]: You need to check the class type first. I presume that’s what the super call did before and why CI is failing.

File: src/twisted/internet/address.py [hunk 3]
```diff
@@ -188,6 +132,17 @@ def __eq__(self, other):
                     if not _PY3 and not platform.isLinux():
                         raise e
             return res
+    else:
+        def __eq__(self, other):
+            if isinstance(other, self.__class__):
```
Comment (hynek) [inline src/twisted/internet/address.py:136]: Basically this.

**Review by hynek — DISMISSED** [review]

File: src/twisted/internet/address.py (hunk 2 above)
Comment (hawkowl) [reply src/twisted/internet/address.py]: fix'd

File: src/twisted/internet/address.py (hunk 1 above)
Comment (hawkowl) [reply src/twisted/internet/address.py]: type is a keyword :)

Comment (hynek) [reply src/twisted/internet/address.py]: Nah, it’s fine:

```
In [1]: import attr

In [2]: @attr.s
   ...: class C:
   ...:     type = attr.ib()
   ...:

In [3]: C(42)
Out[3]: C(type=42)
```

File: src/twisted/internet/address.py [hunk 4]
```diff
@@ -117,66 +92,35 @@ class HostnameAddress(FancyEqMixin, object):
...
-        On UNIX, paths are always bytes. However, as paths are L{unicode} on
-        Python 3, and L{UNIXAddress} technically takes a file path, we convert
-        it to bytes to maintain compatibility with C{os.path} on Python 3.
-        """
-        if name is not None:
-            self._name = _asFilesystemBytes(name)
-        else:
-            self._name = None
-
+    name = attr.ib(converter=lambda x: _asFilesystemBytes(x)
```
Comment (hynek) [inline src/twisted/internet/address.py]: You can simplify this with http://www.attrs.org/en/stable/api.html#attr.converters.optional

**Pull request merged**